kyo-prelude/shared/src/main/scala/kyo2/kernel/ContextEffect.scala

```diff
@@ -0,0 +1,69 @@
+package kyo2.kernel
+
+import internal.*
+import kyo.Tag
+import kyo2.bug
+import kyo2.isNull
+import scala.util.NotGiven
+
+abstract class ContextEffect[+A]
+
+object ContextEffect:
+
+    inline def suspend[A, E <: ContextEffect[A]](inline tag: Tag[E]): A < E =
+        suspend(tag)(identity)
+
+    inline def suspend[A, E <: ContextEffect[A], B, S](
+        inline tag: Tag[E]
+    )(
+        inline f: Safepoint ?=> A => B < S
+    ): B < (E & S) =
+        suspend(tag, bug("Unexpected pending context effect: " + tag.show))(f)
+
+    inline def suspend[A, E <: ContextEffect[A]](
+        inline tag: Tag[E],
+        inline default: => A
+    ): A < Any =
+        suspend(tag, default)(identity)
+
+    inline def suspend[A, E <: ContextEffect[A], B, S](
+        inline _tag: Tag[E],
+        inline default: => A
+    )(
+        inline f: Safepoint ?=> A => B < S
+    )(using inline _frame: Frame): B < S =
+        new KyoDefer[B, S]:
+            def frame = _frame
+            def apply(v: Unit, context: Context)(using Safepoint) =
+                Safepoint.handle(
+                    suspend = this,
+                    continue = f(context.getOrElse(_tag, default).asInstanceOf[A])
+                )
+
+    inline def handle[A, E <: ContextEffect[A], B, S](
+        inline _tag: Tag[E],
+        inline ifUndefined: A,
+        inline ifDefined: A => A
+    )(v: B < (E & S))(
+        using inline _frame: Frame
+    ): B < S =
+        def handleLoop(v: B < (E & S))(using Safepoint): B < S =
+            v match
+                case <(kyo: KyoSuspend[IX, OX, EX, Any, B, S] @unchecked) =>
+                    new KyoSuspend[IX, OX, EX, Any, B, S]:
+                        val tag   = kyo.tag
+                        val input = kyo.input
+                        def frame = _frame
+                        def apply(v: OX[Any], context: Context)(using Safepoint) =
+                            val tag   = _tag
+                            val value = context.getOrElse(tag, null)
```

Comment:
can't this just be `get`? Or is it cheaper to avoid the `bug` function allocation?

Reply:
I was a bit lazy here, the `null` is used to check if the the value is in the context or not. I've exposed a `Context.contains` method to avoid it. About `bug`, in `Context` there's no allocation for it because `default` gets inlined in `getOrElse` directly into the `if` condition body.